### PR TITLE
Rewrite EscapeHTML func to be safer

### PR DESCRIPTION
**Pull request by rougetimelord**
Mozilla kinda had a point, we set the href attribute of an anchor tag to the escaped string. Therefore, we need to make sure to escape single and double quotes.

Definitely could be minimized slightly, but I want to make sure the reviewer understands what's happening fully. This is safe, it destroys tags, and does not allow you to escape the confines of an attribute in order to add tag level event handlers.

See #294 for more details.

1. [x] ensure `src/manifest.ts` and `package.json` have the correct version number
2. [x] use makefile to generate zips (`make chrome`, `make firefox`)
    - [x] chrome should be tested with `npm run build`
    - [ ] test firefox locally using zip
3. [ ] merge to main

### CODE REVIEW

**Comment by cooljeanius** [comment]
I'll leave this one for @kheina to review; don't feel qualified to review it myself

**Comment by bkh555** [comment]
Is there any update on this? Content Security Policies renders my basic XSS attempts useless though I don't see Mozilla budging on this since it's a fairly common issue that has easy fixes.

**Comment by cooljeanius** [comment]
> Is there any update on this? Content Security Policies renders my basic XSS attempts useless though I don't see Mozilla budging on this since it's a fairly common issue that has easy fixes. In fact their doc page states that using innerHTML, like is done on pages/queue/index.ts, is likely to have your add-on rejected. [developer.mozilla.org/en-US/docs/Web/API/Element/innerHTML#security_considerations](https://developer.mozilla.org/en-US/docs/Web/API/Element/innerHTML#security_considerations)

Well do you think that this PR would be an adequate solution for it? I'll approve it if you think it looks good

**Comment by bkh555** [comment]
> Well do you think that this PR would be an adequate solution for it? I'll approve it if you think it looks good

Since Mozilla stated that merely sanitizing the data will satisfy them, I think starting with that is a good idea. 
I'm not a maintainer, just wanted to throw out more information to help get this thing along. I would refer to https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html to be on the safe side.

> OWASP recommends [DOMPurify](https://github.com/cure53/DOMPurify) for HTML Sanitization.

**Comment by rougetimelord** [comment]
Imo it's a non issue... but if we wanted to be extra sure they let us
through, we could use DOMPurify within the modal creation function.

In terms of what this change does, it's basically the same as the old
version but more explicit and removes attribute escapes.

**Comment by cooljeanius** [comment]
> Imo it's a non issue... but if we wanted to be extra sure they let us through, we could use DOMPurify within the modal creation function.

So do you want to do that in this PR, or save it for a separate one?

**Comment by kheina** [comment]
I've held off from pulling in third party libraries to avoid bloating the size of the addon. *if* we rewrite it we should just use the 3-4 line div method they provide.

that being said, I don't want to do either since it's unnecessary, but I will acquiesce if yall would like to change it for firefox approval reasons

**Comment by cooljeanius** [comment]
> I've held off from pulling in third party libraries to avoid bloating the size of the addon. _if_ we rewrite it we should just use the 3-4 line div method they provide.
> 
> that being said, I don't want to do either since it's unnecessary, but I will acquiesce if yall would like to change it for firefox approval reasons

I would like to change it for Firefox approval reasons (assuming that it's correct)

**Comment by bkh555** [comment]
>I would like to change it for Firefox approval reasons (assuming that it's correct)

The fix satisfies the conditions provided by OWASP, so I agree.

>Encoding Type: HTML Entity Encoding Mechanism: Convert & to `&amp;`, Convert < to `&lt;`, Convert > to `&gt;`, Convert " to `&quot;`, Convert ' to `&#x27;`

**Review by cooljeanius — DISMISSED** [review]
approving per @bkh555

**Comment by cooljeanius** [comment]
> 2\. [ ]  use makefile to generate zips (`make chrome`, `make firefox`)

...actually wait, I'm getting some errors when I try to do this step:
```
$ make chrome
cat: build/manifest.json: No such file or directory
makefile:6: *** Extension version mismatch. manifest: , package.json: "0.4.3".  Stop.
$ make firefox
cat: build/manifest.json: No such file or directory
makefile:6: *** Extension version mismatch. manifest: , package.json: "0.4.3".  Stop.
```

File: src/utilities.ts
Comment (kheina) [inline src/utilities.ts:203]: any chance we can do something like this rather than a bunch of chained replaces that all need to parse the full string? 

https://github.com/kheina-com/frontend/blob/37f3a35ae41d7639d405e55e9a4a8c12526fcccc/src/utilities/markdown.js#L142-L148

I'm also not 100% sure what this is doing or why it's necessary `.replace(/&(?![#\w]{2,};)/g, '&amp;')` I don't think there's anything necessarily risky about ampersands in particular, but I'm not against escaping them, just in case.

I would also just return here rather than storing it in a variable and then escaping afterwards

**Comment by cooljeanius** [comment]
> > 2. [ ]  use makefile to generate zips (`make chrome`, `make firefox`)
> 
> ...actually wait, I'm getting some errors when I try to do this step:
> 
> ```
> $ make chrome
> cat: build/manifest.json: No such file or directory
> makefile:6: *** Extension version mismatch. manifest: , package.json: "0.4.3".  Stop.
> $ make firefox
> cat: build/manifest.json: No such file or directory
> makefile:6: *** Extension version mismatch. manifest: , package.json: "0.4.3".  Stop.
> ```

This still happens.

**Comment by kheina** [comment]
> > > 2. [ ]   use makefile to generate zips (`make chrome`, `make firefox`)
> > 
> > 
> > ...actually wait, I'm getting some errors when I try to do this step:
> > ```
> > $ make chrome
> > cat: build/manifest.json: No such file or directory
> > makefile:6: *** Extension version mismatch. manifest: , package.json: "0.4.3".  Stop.
> > $ make firefox
> > cat: build/manifest.json: No such file or directory
> > makefile:6: *** Extension version mismatch. manifest: , package.json: "0.4.3".  Stop.
> > ```
> 
> This still happens; is that expected?

I'll take a look after work

**Comment by kheina** [comment]
> > > 2. [ ]   use makefile to generate zips (`make chrome`, `make firefox`)
> > 
> > 
> > ...actually wait, I'm getting some errors when I try to do this step:
> > ```
> > $ make chrome
> > cat: build/manifest.json: No such file or directory
> > makefile:6: *** Extension version mismatch. manifest: , package.json: "0.4.3".  Stop.
> > $ make firefox
> > cat: build/manifest.json: No such file or directory
> > makefile:6: *** Extension version mismatch. manifest: , package.json: "0.4.3".  Stop.
> > ```
> 
> This still happens; is that expected?

this is not expected, are you running `npm run build` first?

**Comment by cooljeanius** [comment]
> > > > 2. [ ]    use makefile to generate zips (`make chrome`, `make firefox`)
> > > 
> > > 
> > > ...actually wait, I'm getting some errors when I try to do this step:
> > > ```
> > > $ make chrome
> > > cat: build/manifest.json: No such file or directory
> > > makefile:6: *** Extension version mismatch. manifest: , package.json: "0.4.3".  Stop.
> > > $ make firefox
> > > cat: build/manifest.json: No such file or directory
> > > makefile:6: *** Extension version mismatch. manifest: , package.json: "0.4.3".  Stop.
> > > ```
> > 
> > 
> > This still happens; is that expected?
> 
> this is not expected, are you running `npm run build` first?

Ah ok, now it works...

**Review by cooljeanius — APPROVED** [review]

**Comment by cooljeanius** [comment]
Zips from this PR:
[blue-blocker-chrome-0.4.3.zip](https://github.com/user-attachments/files/16118166/blue-blocker-chrome-0.4.3.zip)
[blue-blocker-firefox-0.4.3.zip](https://github.com/user-attachments/files/16118167/blue-blocker-firefox-0.4.3.zip)

File: src/utilities.ts
Comment (cooljeanius) [reply src/utilities.ts:203]: is this a blocker for merging, or can it be fixed later?

Comment (rougetimelord) [reply src/utilities.ts:203]: > I'm also not 100% sure what this is doing or why it's necessary `.replace(/&(?![#\w]{2,};)/g, '&amp;')`

If you're confused why it's so complicated, it's so that the replacement doesn't take something like `&gt;` and spit out `&amp;gt;`. I'm ngl I also don't get what the threat with ampersands is, but [OSWAP](https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html#output-encoding-rules-summary) says to replace it with an HTML entity.

Comment (bkh555) [reply src/utilities.ts:203]: <, >, and & are already sanitized to when you call 'const partiallySafe = element.innerHTML;' so those parts are unnecessary IMO 
image

Comment (bkh555) [reply src/utilities.ts:203]: As to why ampersands are escaped: https://stackoverflow.com/questions/39214564/why-ampersand-should-be-escaped-because-of-xss-injection

Comment (rougetimelord) [reply src/utilities.ts:203]: > <, >, and & are already sanitized to when you call 'const partiallySafe = element.innerHTML;' so those parts are unnecessary IMO image

They are, I mostly just added it to the replace chain so that Mozilla's reviewers can see the replacement happening explicitly? Might not be necessary?

Comment (rougetimelord) [reply src/utilities.ts]: Ok removed the explicit <, >, and & replacements. I also validated that the OSWAP 5 are in fact replaced 
image

**Comment by rougetimelord** [comment]
Going ahead with the merge so we can hopefully get 0.4.3 out soon 🤞🏻

**Pull request merged**
